python/kserve/kserve/inference_client.py

```diff
@@ -12,43 +12,150 @@
+    :param private_key: (optional) Path to the PEM-encoded private key file as a string or None if no private key
+                        should be used. If creds provided this will be ignored.
+    :param certificate_chain: (optional) Path to the PEM-encoded certificate chain file as a string or None if no
+                              certificate chain should be used. If creds provided this will be ignored.
+    :param creds: (optional) A ChannelCredentials instance for secure channel communication.
+    :param channel_args: (optional) An list of key-value pairs (channel_arguments in gRPC Core runtime) to configure
+                         the channel.
+    :param timeout (optional) The maximum end-to-end time, in seconds, the request is allowed to take. By default,
+                   client timeout is 60 seconds. To disable timeout explicitly set it to 'None'.
+    """
+
     def __init__(
         self,
-        url,
-        verbose=False,
-        ssl=False,
-        root_certificates=None,
-        private_key=None,
-        certificate_chain=None,
-        creds=None,
-        channel_args=None,
+        url: str,
+        verbose: bool = False,
+        use_ssl: bool = False,
+        root_certificates: str = None,
+        private_key: str = None,
+        certificate_chain: str = None,
+        creds: grpc.ChannelCredentials = None,
+        channel_args: List[Tuple[str, Any]] = None,
+        timeout: Optional[float] = 60,
     ):
 
+        # requires appending the port to the predictor host for gRPC to work
+        if ":" not in url:
+            port = 443 if use_ssl else 80
+            url = f"{url}:{port}"
+        # Default retry config
+        service_config_json = json.dumps(
+            {
+                "methodConfig": [
+                    {
+                        # Apply retry to all methods
+                        "name": [{}],
+                        "retryPolicy": {
+                            "maxAttempts": 3,
+                            "initialBackoff": "0.1s",
+                            "maxBackoff": "1s",
+                            "backoffMultiplier": 2,
+                            "retryableStatusCodes": ["UNAVAILABLE"],
+                        },
+                    }
+                ]
+            }
+        )
         # Explicitly check "is not None" here to support passing an empty
         # list to specify setting no channel arguments.
         if channel_args is not None:
             channel_opt = channel_args
+            if ("grpc.enable_retries", 0) not in channel_opt:
+                is_exist = False
+                for key, _ in channel_opt:
+                    if key == "grpc.service_config":
+                        is_exist = True
+                        break
+                if not is_exist:
+                    channel_opt.append(("grpc.enable_retries", 1))
+                    channel_opt.append(("grpc.service_config", service_config_json))
```

Comment:
if `("grpc.enable_retries", 1)` exists in the channel args then it we append a duplicated entry ?

Reply:
Fixed.